core/src/layers/prometheus_client.rs

```diff
@@ -159,32 +162,78 @@ impl PrometheusClientMetrics {
             bytes_total,
         }
     }
+}
+
+#[derive(Clone, Debug)]
+struct PrometheusClientMetrics {
+    metrics: Arc<PrometheusClientMetricDefinitions>,
+    scheme: Scheme,
+    root: String,
+    name: String,
+}
+
+impl PrometheusClientMetrics {
+    fn new(
+        metrics: Arc<PrometheusClientMetricDefinitions>,
+        scheme: Scheme,
+        root: String,
+        name: String,
+    ) -> Self {
+        Self {
+            metrics,
+            scheme,
+            root,
+            name,
+        }
+    }
 
-    fn increment_errors_total(&self, scheme: Scheme, op: &'static str, err: ErrorKind) {
+    fn increment_errors_total(&self, op: &'static str, err: ErrorKind) {
```

Comment:
I'm thinking of implementing `EncodeLabelSet` directly for our lables to avoid `to_string()` here.

ref: https://docs.rs/prometheus-client/latest/prometheus_client/metrics/family/struct.Family.html#family-with-custom-type-for-performance-andor-type-safety

Would you like to start a new PR for this or just in this PR?

Reply:
let me have a try in this pr

Reply:
i switched to a `Cow<'static, str>` in https://github.com/apache/opendal/pull/4907/commits/daab9db22bb84e25f959a44bb0813c44ae32a504 which helps reduces copy but can not reduce all the copies.

it turns out that we can not simply add lifetime annotations (for the labels like `root` and `namespace` which stored in a singleton struct) in the labels type with promtheus-client, because each metric (with labels together) will be registered into a global registry, which requires the metrics to be `'static`.

Reply:
Hi, I'm thinking about making `OperationLabels` a seperate struct and implement `EncodeLabelSet` for it. For example:

```rust
struct OperationLabels {
  scheme: Scheme,
  op: &'static str,
  namespace: Arc<String>,
  root: Arc<String>,
}

impl EncodeLabelSet for OperationLabels { .. }
```

In this way, we can constrct `OperationLabels` and use:

```rust
family.get_or_create(&OperationLabels { ... }).inc();
```

Reply:
makes sense, i've changed them to structs, PTAL